src/backend/cuda/kernel/join.hpp

```diff
@@ -73,6 +73,15 @@ namespace cuda
                         blocksPerMatY * X.dims[3],
                         1);
 
+            int maxBlocksY = cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize[1];
+            // Increase the work per block until the condition is satisfied
+            // Increase threads only upto 512
+            while(blocks.y > maxBlocksY) {
+                if(threads.x * threads.y < 512) threads.y *= 2;
```

Comment:
Yes, that is the basis of this kind of loop. It assumes that the kernels operate in a specific manner for X and Y.
The X thread/blocks will operate on 0 and 2 dims, Y operates on 1 and 3 with threads doing multiple elements. Only those kind of kernels can use this method of reducing no. of blocks. So once we hit 512 threads, we simply increase the number of iterations per thread.
It's definitely not universal.